vortex-file/src/read/coalescer.rs

```diff
@@ -0,0 +1,220 @@
+                            }
+                        }
+                    } else {
+                        *queued_res = RowMaskState::Empty;
+                    }
+                }
+                RowMaskState::Ready(_) => {}
+                RowMaskState::Empty => {}
+            }
+        }
+
+        while read_more_count < NUM_TO_COALESCE {
+            if let Some(next_s) = ready!(self.values_iter.poll_next_unpin(cx)) {
+                let next_mask = next_s?;
+                if let Some(read_result) = self.row_mask_reader.read_mask(&next_mask)? {
+                    match read_result {
+                        MessageRead::ReadMore(m) => {
+                            self.queued.push_back(RowMaskState::Pending(next_mask));
+                            to_read.extend(m);
+                            read_more_count += 1;
+                        }
+                        MessageRead::Value(v) => {
+                            self.queued.push_back(RowMaskState::Ready(v));
+                        }
+                    }
+                }
+            } else {
+                break;
+            }
+        }
+        Poll::Ready(Ok(to_read))
+    }
+
+    fn dispatch_messages(
+        &self,
+        messages: Vec<MessageLocator>,
+        waker: Waker,
+    ) -> BoxFuture<'static, io::Result<Vec<Message>>> {
+        let reader = self.io_read.clone();
+        self.dispatcher
+            .dispatch(move || async move {
+                let read_messages = reader
+                    .read_byte_ranges(messages.iter().map(|msg| msg.1.to_range()).collect())
+                    .map(move |read_res| {
+                        Ok(messages
+                            .into_iter()
+                            .map(|loc| loc.0)
+                            .zip(read_res?)
+                            .map(|(loc, bytes)| Message(loc, bytes))
+                            .collect())
+                    })
+                    .await;
+                waker.wake();
+                read_messages
+            })
+            .vortex_expect("Async task dispatch")
+            .map(|res| res.unwrap_or_else(|e| Err(io::Error::new(ErrorKind::Other, e))))
+            .boxed()
+    }
+}
+
+impl<R, S, V, RM> Stream for Coalescer<R, S, V, RM>
+where
+    R: VortexReadAt + Unpin,
+    S: Stream<Item = VortexResult<RowMask>> + Unpin,
+    RM: RowMaskReader<V> + Unpin,
+    V: Unpin,
+{
+    type Item = VortexResult<V>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        if let Some(in_flight) = &mut self.in_flight {
+            match in_flight.poll_unpin(cx) {
+                Poll::Ready(msgs) => {
+                    self.store_messages(
+                        msgs.map_err(|e| vortex_err!("Cancelled in flight read {e}"))?,
+                    );
+                    let messages = ready!(self.gather_read_messages(cx))?;
```

Comment:
If this returns pending we get into an error state where we poll a done future which in rust is a bug. Need this to not propagate pending